### ISSUE TITLE
With Multiple requests fails

### ISSUE THREAD

**Issue by elreyponce** [HakanL/WkHtmlToPdf-DotNet#124]
Using .NET API and sending multiple requests at the same time I got this error

![imagen](https://github.com/HakanL/WkHtmlToPdf-DotNet/assets/31009526/2f39afe9-c1a4-4df2-84ab-f275794f9b76)

**Comment by HakanL** [comment]
Do you have multiple instances of the `SynchronizedConverter`? You need to just have one, as a singleton.

**Comment by elreyponce** [comment]
I have only one instance injected as Singleton in my API

**Comment by HakanL** [comment]
Sounds like there are multiple instances somehow. I can't see what the actual exception is, but if you can show it in English and post your code where you call the convert and wire up the instance then we can see if we spot anything.

**Comment by elreyponce** [comment]
This is the information from a DMP file which I extracted from my application pool windows process.
![imagen](https://github.com/HakanL/WkHtmlToPdf-DotNet/assets/31009526/cdc6eb8b-20c3-435d-8bf1-4da91f432efa)

I've tried blocking the concurrency using SemaphoreSlim before the "convert" function but I had the same error.

Here's part of my code.
1. This is how I declare IConverter as a singleton intance.
![imagen](https://github.com/HakanL/WkHtmlToPdf-DotNet/assets/31009526/1c1c8f1a-bdcd-4ce2-9ac6-499f87ea921c)

3. Injecting the IConverter interface.
![imagen](https://github.com/HakanL/WkHtmlToPdf-DotNet/assets/31009526/b9d7ef61-385d-4289-9bec-dcd1409aa569)

4. This is how I invoke the PDF generation.
![imagen](https://github.com/HakanL/WkHtmlToPdf-DotNet/assets/31009526/afd2795d-2ab3-41fd-9056-4a8537b360e6)

**Comment by HakanL** [comment]
Your code looks correct and solid. It sounds like IIS (or .NET hosting, not sure) has a limit on how long the requests can take (and PDF generation takes a while) and when you have multiple concurrent requests then you'll be over the limit (the WkHtmlToPdf project only allows a single concurrent request, the more requests you have the longer each thread will wait). I typically have a queue for processing long-running jobs and not have the web server hang for the result.

**Comment by elreyponce** [comment]
Thank for your quick responses, and for your work with this library. 

Is it difficult to change the "Convert" method to an asynchronous one? That is the suggestion.
![imagen](https://github.com/HakanL/WkHtmlToPdf-DotNet/assets/31009526/c43f0ed0-90f0-4279-b55d-2543032e82e6)

**Comment by HakanL** [comment]
Probably not possible, this project is just a wrapper for the native library that does the actual work, and it's unfortunately only single-threaded. You could potentially load multiple add domains and load the library in each one, but I've never done anything like this so I'm not sure how to do it, or if it would help, it's just a thought.

**Comment by elreyponce** [comment]
Ok, I'll try it. Thank you!!!

**Comment by mwwhited** [comment]
If you end up with multiple instances of the IOC container `SynchronizedConverter` will not be safe enough.  Here is a fix that was able to ensure a real single instance even across multiple IServiceProvider instances

```
    private static IConverter? _converter; 

    public static IServiceCollection TryAddWkHtmlToPdfServices(this IServiceCollection services)
    {
        services.AddTransient<ITools, PdfTools>();
        // Another option is to change the internal fields in the SynchronizedConverter into static fields instead of instance fields
        services.AddSingleton<IConverter>(sp => _converter ??= ActivatorUtilities.CreateInstance<SynchronizedConverter>(sp));
        return services;
    }
    ```

**Comment by mwwhited** [comment]
If you end up with multiple instances of the IOC container `SynchronizedConverter` will not be safe enough.  Here is a fix that was able to ensure a real single instance even across multiple IServiceProvider instances

```
    private static IConverter? _converter; 

    public static IServiceCollection TryAddWkHtmlToPdfServices(this IServiceCollection services)
    {
        services.AddTransient<ITools, PdfTools>();
        // Another option is to change the internal fields in the SynchronizedConverter into static fields instead of instance fields
        services.AddSingleton<IConverter>(sp => _converter ??= ActivatorUtilities.CreateInstance<SynchronizedConverter>(sp));
        return services;
    }
    ```

**Comment by HakanL** [comment]
@mwwhited Hmm, why would you have multiple service providers, that sounds risky? But if that's the case then your code would work, the native library only supports single-threaded access, so it has to be a maximum of 1 library instance per AppDomain.

**Comment by mwwhited** [comment]
Primarily due to mstest and how it parallels applications since they are not process isolated.  It would be possible is using middleware isolation within and application.  Unlikely for this particular code but is possible.  A better model for this framework would be an operating system level semaphore if you are saying there can only be one instance at a time.

Either way my IoC registration resolve the deadlock issue if anyone else is effected.

**Comment by HakanL** [comment]
I reviewed this again. It seems that you're creating a new task and then waiting on that. That's typically not a great way to do it, I believe you can get some deadlocks and issues. I would recommend that you just call it synchronously instead, adding it to a thread pool if necessary to run it in the background.

**Comment by elreyponce** [comment]
Hi @HakanL, thanks for your response. I did some tests with IIS and with Kestrel servers. With Kestrel the process never blocked when I sent many simultaneous requests. I had to do a redirection from the IIS server to the Kestrel server in order to solve it.

**Comment by HakanL** [comment]
@elreyponce I think IIS may spin up multiple processes somehow, which causes the issue. I know IIS does some special tricks in some scenarios.
